app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/DeviceListFragment.java

```diff
@@ -0,0 +1,357 @@
+import androidx.fragment.app.ListFragment;
+
+import org.kiwix.kiwixmobile.BuildConfig;
+import org.kiwix.kiwixmobile.R;
+import org.kiwix.kiwixmobile.utils.SharedPreferenceUtil;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.InetAddress;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.kiwix.kiwixmobile.zim_manager.local_file_transfer.LocalFileTransferActivity.showToast;
+
+/**
+ * Part of the local file sharing module, this fragment is responsible for displaying details of
+ * the user device and the list of available peer devices, and once the user taps on a particular
+ * peer device item, connecting to that device and initiating file transfer.
+ *
+ * File transfer involves two phases:
+ * 1) Handshake with the selected peer device, using {@link PeerGroupHandshakeAsyncTask}
+ * 2) After handshake, starting the files transfer using {@link SenderDeviceAsyncTask} on the sender
+ *    device and {@link ReceiverDeviceAsyncTask} files receiving device
+ *
+ * The starting point for the module is {@link LocalFileTransferActivity}
+ * */
+public class DeviceListFragment extends ListFragment implements WifiP2pManager.PeerListListener, WifiP2pManager.ConnectionInfoListener {
+
+  public static final String TAG = "DeviceListFragment";
+  public static int PEER_HANDSHAKE_PORT = 8009;
+  public static int FILE_TRANSFER_PORT = 8008;
+
+  private SharedPreferenceUtil sharedPreferenceUtil;
+
+  private LocalFileTransferActivity localFileTransferActivity;  // Parent activity, starting point of the module
+  private TransferProgressFragment transferProgressFragment;    // Sibling fragment, for displaying transfer progress
+
+  private View deviceListFragmentRootView = null; // Root view corresponding to the DeviceListFragment
+
+  private boolean fileSender = false; // Whether file sending device or not
+  private ArrayList<Uri> fileUriList; // Uris of files to be shared, available only for the sender device
+  private int totalFilesForTransfer = -1;
+  private int filesSent = 0;          // Count of number of files transferred until now
+  private ArrayList<FileItem> filesToSend = new ArrayList<>();
+
+  private WifiP2pDevice userDevice;   // Represents the device on which the app is running
+  private WifiP2pInfo groupInfo;      // Corresponds to the WiFi P2P group formed between the two devices
+  private List<WifiP2pDevice> peerDevices = new ArrayList<WifiP2pDevice>();
+
+  private WifiP2pDevice selectedPeerDevice = null;
+  private InetAddress selectedPeerDeviceInetAddress;
+
+  private InetAddress fileReceiverDeviceAddress;  // IP address of the file receiving device
+  private boolean fileTransferStarted = false;
+
+  @Override
+  public void onActivityCreated(@Nullable Bundle savedInstanceState) {
+    super.onActivityCreated(savedInstanceState);
+    localFileTransferActivity = (LocalFileTransferActivity) getActivity();
```

Comment:
I am not a fan of this tight coupling to LocalFileTransferActivity. General practice is to communnicate through an interface and to cast the context to the interface in `onAttach`, that way theoretically we can use this with any activity

Reply:
DeviceListFragment is statically added when LocalFileTransferActivity is created, so I felt tight-coupling was going to be unavoidable, which is why I added all the files to a sub-package

What should I do here?

Reply:
I am switching over to your branch to get a better grasp of how this all works.
I suspect that significant rework is needed here, perhaps custom views to encapsulate logic to a specific view portion, much less annoying and complex than fragments.
There are problems with data flow as it seems everything needs to know about everything else.
We should streamline this as much as possible, thinking about actions performable by the user and the states the UI can be in

Reply:
I think step 1 is 2 interfaces, FileReceiver & FileSender and 2 concrete instances WifiP2PFileSender/Receiver.
We can pass callbacks in here for actions we want to take on the events we care about onError/onFilesReceived/OnProgress.
There is also concern around using AsyncTask, they are much maligned in the android community over how easy it is to leak context with them. Rxjava and wrapping all disposables in a compositeDisposable is generally preferred though coroutines are gaining popularity however both of these are complex topics.

Reply:
>There are problems with data flow as it seems everything needs to know about everything else.

An overview of what's going on:
So there's a `WifiP2pManager` which resides in an activity (_LocalFileTransferActivity_ here) and handles everything to do with the device's wifi-p2p framework. So anything to do with the wifi connection requires the activity.

_DeviceListFragment_ handles display of available peers (obtained from the _WifiP2pManager_), then the selection of a peer by the user, followed by starting the file transfer.
It initially obtains: 1) info whether the device is a sender or receiver, and 2) list of file URIs to be sent (in case of a sender); from the _LocalFileTransferActivity_. Subsequent interaction is only to interact with the p2p-manager or to close the activity post transfer.

It also starts an _async-task_ on both the devices for a handshake between the two (called _PeerGroupHandshakeAsyncTask_), followed by async-tasks for file transfer (called _Sender/ReceiverDeviceAsyncTask_ ). The sender async-task obtains the list of file uris from the _DeviceListFragment_ while the receiver device gets the file details from the handshake. Subsequently, they communicate with the _TransferProgressFragment_ to update status of files being transferred. At the end of transfer, they communicate with the _LocalFileTransferActivity_ to close it.

@macgills what all should have been avoided here?

Reply:
> I suspect that significant rework is needed here, perhaps custom views to encapsulate logic to a specific view portion

I'll need guidance with this as to what is to be done. (Also I haven't created custom views previously)

> I think step 1 is 2 interfaces, FileReceiver & FileSender and 2 concrete instances WifiP2PFileSender/Receiver.

I'm not clear how they will be structured
Will `WifiP2PFileSender/Receiver` extend the interfaces? Because the only thing common to both sender and receiver which comes to my mind is the WifiP2pManager, which is sort of already obtained from the common source of LocalFileTransferActivity

Reply:
> There is also concern around using AsyncTask, they are much maligned in the android community over how easy it is to leak context with them. Rxjava and wrapping all disposables in a compositeDisposable is generally preferred though coroutines are gaining popularity however both of these are complex topics.

Maybe I can add a TODO for now and work on it later? I certainly plan to learn RxJava, and being able to improve this will be an added advantage

Reply:
Okay so now that we are through phase 1 I have a clearer view on this.
I think all the fragments are just making the code more complex right now and we will never really want to reuse them and there is no real demand from android that you have them so DeviceList & TransferProgress can be integrated into the LocalFileTransfer view heirarchy.

Unfortunately this bloats LocalFileTransfer a lot so that must be addressed. I want to see an object that encapsulates all of WifiP2P, something like WifiDirectManager in LocalTransferActivity. It will remove most of these static methods I imagine and the only thing that LocalTransferActivity should be doing is setting the visibility of views, setting data on views and forwarding user actions to WifiDirectManager. We need to make LocalTransferActivity as stateless as possible is the overall aim.

If you need any guidance get me on slack as it will be much quicker than commenting on this.

Reply:
[Last commit](https://github.com/kiwix/kiwix-android/pull/1276/commits/709e043f68b9771ca90a8d5dc5b43ceee27928d4) of the restructuring